dev_tools/oneapi/fake_interface/cooperative_groups.cuh

```diff
@@ -0,0 +1,57 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_FAKE_INTERFACE_COOPERATIVE_GROUPS_CUH_
+#define GKO_FAKE_INTERFACE_COOPERATIVE_GROUPS_CUH_
+
+#include <cuda/components/cooperative_groups.cuh>
```

Comment:
In general, yes.
we have cuda code and corresponding dpcpp code but dpct does not allow us to provide custom mapping.
This file only involves dpct process such that dpct has the information on interface but does not try converting real cuda code. In the end, we replace the fake interface to our real dpcpp impl

Reply:
But `cuda/components/cooperative_groups.cuh` has cuda syntax, right ? Do you convert/replace that or dpct does that ?

Reply:
dpct has different behavior on system header and local header. moreover, they use keyword to capture cuda function.
when function is in local header, dpct will try to convert them.
but in system header, dpct will only consider it is a function without checking anything.
cooperative group will in their keyword list, so it throws error.
we put the real cuda cooperative group as system header to avoid dpct touching and fake_interface with alternative name to avoid dpct detect.
we also add threadIdx in the function such that dpct can add the nd_item correctly.
in dpct view, it is a normal function with a system header function and threadIdx.
Thus, dpct does not really see cuda/components/cooperative_groups.cuh.
and porting script will replace fake_interface with corresponding dpcpp impl

Reply:
I see. Thanks for the explanation.